design/design-doc-agent.md

```diff
@@ -0,0 +1,627 @@
+team, which is `*Requirement` types and consume the infrastructure by mounting
+the secret whose name they specified on the `*Requirement` custom resource. To
+keep this experience, we need to have a synchronization loop for the following
+resources:
+
+* Pull
+  * `CustomResourceDefinition`s of all types that we want the applications to be
+    able to manage and view:
+    * Requirements that are published via `InfrastructurePublication`s. The
+      source of truth will be the remote cluster.
+    * `InfrastructureDefinition`, `InfrastructurePublication` and `Composition`.
+  * `Composition`s to discover what's available to choose. CRs of this type will
+    be read-only, and the source of truth will be the central cluster.
+  * `InfrastructurePublication`s to discover what's availabe as published API.
+    Read-only.
+  * `InfrastructureDefinition`s to discover how the secret keys are shaped.
+    Read-only.
+  * `Secret`s that are result of the infrastructure that is provisioned so that
+    it can be mounted to `Pod`s. Read-only.
+* Push
+  * `*Requirement` custom resources so that infrastructure can be requested.
+    Read and write permissions in a specific namespace in the central cluster
+    will be needed.
+
+### RBAC
+
+As we have two different Kubernetes clusters, there will be two separate security
+domains and because of that, the `ServiceAccount`s in the remote cluster will
+not be able to do any operation in the central cluster. Since the entity that
+will execute the operations in the central cluster is the agent, we need to
+define how we can deliver the necessary credentials to the agent so that it can
+connect to the central cluster. Additionally, it will need some permissions to
+execute operations in the remote cluster like `Secret` and
+`CustomResourceDefinition` creation. We will look at how the agent will be
+granted permissions to do its job in two separate domains with different
+mechanisms.
+
+#### Authenticating to The Central Cluster
+
+In order to execute any operation, a `ServiceAccount` needs to exist in the
+central cluster with appropriate permissions to read during pull and write
+during push operations while synchronizing with central cluster. Since the agent
+is running in the remote cluster, the credentials of this `ServiceAccount` will
+be stored in a `Secret` in the remote cluster. Alongside the credentials, the
+agent needs to know the namespace that it should sync to in the central cluster.
+
+The easiest configuration would be the one where we specify the `Secret` and a
+target namespace via installation commands of the agent. However, the namespace
+pairing list could change dynamically after the installation. So, we need to be
+able to supply the pairings after the installation. In order to specify which
+namespace in the remote cluster should be synced to which namespace in the
+central cluster, `Namespace` resource needs to be annotated.
+
+```yaml
```

Comment:
is this installed on the control plane cluster or the remote cluster?

Reply:
@muvaf should confirm, but we chatted earlier today and my understanding is that the annotation would be on the namespace on the remote (agent side) cluster.

Reply:
Yes, this is in the remote cluster.